In-App Updates: Prompt update after phased release is complete

momo-ozawa opened the pull request:
Part of https://github.com/Automattic/wordpress-mobile/issues/55
Part of https://github.com/Automattic/wordpress-mobile/issues/56

## Description
- Adds a check to only show the update prompts if it's been 7 days since the latest version was released (i.e. [phased release](https://developer.apple.com/help/app-store-connect/update-your-app/release-a-version-update-in-phases/) is at 100%), to prevent showing the prompt to users who have automatic updated enabled.

## How to test

### Flexible update

**Preconditions**
- In Xcode, change the app version to something lower than the current app store version, e.g. 24.6
- Enable the `In-App Updates` remote feature flag

**Test 1.1**
- Run on a real device
- ✅ Verify the flexible update is NOT displayed

**Test 1.2**
- Change `delayInDays` default value to 1
- Run on a real device
- ✅ Verify the flexible update is displayed

### Blocking update

**Preconditions**
- In Xcode, change the app version to something lower than the current app store version, e.g. 24.6
- Enable the `In-App Updates` remote feature flag
- Change the `In-App Update Blocking Version` remote config value to 24.7

**Test 2.1**
- Run on a real device
- ✅ Verify the blocking update is NOT displayed

**Test 2.2**
- Change `delayInDays` default value to 1
- Run on a real device
- ✅ Verify the blocking update is displayed


## Regression Notes
1. Potential unintended areas of impact
Showing app update prompt

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Added unit test

3. What automated tests I added (or what prevented me from doing so)
AppUpdateCoordinatorTests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Add currentVersionReleaseDate

Commit:
Show prompt if it’s been a week since the latest release

Commit:
Update tests

Commit:
Fix whitespace

kean commented on WordPress/Classes/Services/AppUpdate/AppUpdateCoordinator.swift:
```diff
@@ -24,7 +25,8 @@ final class AppUpdateCoordinator {
         isLoggedIn: Bool = AccountHelper.isLoggedIn,
-        isInAppUpdatesEnabled: Bool = RemoteFeatureFlag.inAppUpdates.enabled()
+        isInAppUpdatesEnabled: Bool = RemoteFeatureFlag.inAppUpdates.enabled(),
+        delayInDays: Int = 7
```
Can we delay it by more days, just to be safe? I'd probably go with a month to let users a chance to update manually and to make sure the automatic update has enough time to run. I don't think we need to start bombarding them with the updates right away, do we?

momo-ozawa replied:
This value doesn't represent how many days to wait in between showing prompts.

It represents how many days to wait until a version has been released to show a prompt to update. I chose 7 days because phased release is complete after 7 days. This delay ensures that the prompt isn't shown during the phased release period - it prevents showing the prompt to users who have automatic updates enabled.

As discussed in p1716217030223529-slack-C072JBZL84U, my plan is change the flexible update interval remote config value from 5 to 90 days.

kean commented:
I hit an assertion here:

<img width="430" alt="Screenshot 2024-05-20 at 10 25 24 AM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/b7257e9c-615e-4bf4-a779-9c11dab26e68">

`updateRemoteConfig` is called when the app enters foreground, and it performs a version check every time.

kean commented on WordPress/Classes/Services/AppUpdate/AppStoreSearchService.swift:
```diff
@@ -10,6 +10,13 @@ struct AppStoreLookupResponse: Decodable {
+        func currentVersionHasBeenReleased(for days: Int) -> Bool {
+            let secondsInDay: TimeInterval = 86_400
+            let secondsSinceRelease = -currentVersionReleaseDate.timeIntervalSinceNow
+            return secondsSinceRelease > Double(days) * secondsInDay
```
(nit) there is a convenience API for this:

```
guard let days = Calendar.current.dateComponents([.day], from: lhs, to: rhs).day else {
            return wpAssertionFailure("should-never-happen")
}
return days > 1
```

**warning:** I have not tested it!

momo-ozawa replied:
I've addressed this in https://github.com/wordpress-mobile/WordPress-iOS/pull/23221 🙇‍♀️

momo-ozawa commented:
> updateRemoteConfig is called when the app enters foreground, and it performs a version check every time.

@kean I've addressed this in https://github.com/wordpress-mobile/WordPress-iOS/pull/23221.

I changed it to fetching the app store info once a day.
